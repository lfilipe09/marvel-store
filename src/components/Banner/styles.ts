import styled, { css } from 'styled-components'
import media from 'styled-media-query'

export const Wrapper = styled.main`
  display: flex;
  flex-direction: column-reverse;
  ${media.greaterThan('medium')`
    min-height: 40rem;
    display: grid;
    grid-template-columns: 1fr 1.5fr;
  `}
`

export const Caption = styled.div`
  ${({ theme }) => css`
    width: 100%;
    background-color: ${theme.colors.black};
    padding: ${theme.spacings.small};
    display: flex;
    flex-direction: column;
    grid-gap: ${theme.spacings.xsmall};
    ${media.greaterThan('medium')`
      padding: ${theme.spacings.large};
      grid-gap: ${theme.spacings.small};
      justify-content: center;
    `}
  `}
`

export const Title = styled.h2`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.xlarge};
    font-weight: ${theme.font.bold};
    color: ${theme.colors.white};
    ${media.greaterThan('medium')`
      font-size: 3rem;
      line-height: 3.3rem;
    `}
  `}
`

export const Subtitle = styled.h3`
  ${({ theme }) => css`
    color: ${theme.colors.white};
    font-size: ${theme.font.sizes.small};
    font-weight: ${theme.font.normal};
    margin-bottom: ${theme.spacings.xsmall};
    strong {
      color: ${theme.colors.primary};
      font-weight: ${theme.font.bold};
    }
    ${media.greaterThan('medium')`
      font-size: ${theme.font.sizes.large};
    `}
  `}
`

export type ImageContainerProps = { imgUrl: string }

export const ImageContainer = styled.div<ImageContainerProps>`
  ${({ imgUrl }) => css`
    min-height: 25rem;
    background-image: url(${imgUrl});
    background-size: cover;
    background-repeat: no-repeat;
  `}
`
