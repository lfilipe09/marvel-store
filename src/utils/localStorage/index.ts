import { Auth } from 'types/authTypes'
import { User } from 'types/userTypes'

const APP_KEY = '@bornlogic'

export function getStorageItem(key: string) {
  if (typeof window === 'undefined') return

  const data = window.localStorage.getItem(`${APP_KEY}_${key}`)
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return JSON.parse(data!)
}

export function setStorageItem(key: string, value: Auth | User[]) {
  if (typeof window === 'undefined') return

  const data = JSON.stringify(value)
  return window.localStorage.setItem(`${APP_KEY}_${key}`, data)
}

export function removeStorageItem(key: string) {
  if (typeof window === 'undefined') return

  window.localStorage.removeItem(`${APP_KEY}_${key}`)
}
