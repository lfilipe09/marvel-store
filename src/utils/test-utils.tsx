import React from 'react'
import { ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { AuthContext, AuthInitialValues, AuthProps } from '../hooks/useAuth'

import { ThemeProvider } from 'styled-components'
import theme from '../styles/themes/red'

type CustomRenderProps = {
  cartProviderProps?: AuthProps
} & Omit<RenderOptions, 'queries'>

const customRender = (
  ui: ReactElement,
  {
    cartProviderProps = AuthInitialValues,
    ...renderOptions
  }: CustomRenderProps = {}
) =>
  render(
    <ThemeProvider theme={theme}>
      <AuthContext.Provider value={cartProviderProps}>
        {ui}
      </AuthContext.Provider>
    </ThemeProvider>,
    renderOptions
  )

export * from '@testing-library/react'
export { customRender as render }
