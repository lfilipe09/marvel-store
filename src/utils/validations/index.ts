import Joi from 'joi'
import { UserForm, UserFormSignIn } from '../../types/userTypes'
import {
  cpf_schema,
  email_schema,
  name_schema,
  password_schema,
  surname_schema
} from './errors'

const fieldsValidations = {
  cpf: cpf_schema,
  email: email_schema,
  name: name_schema,
  password: password_schema,
  surname: surname_schema
}

export type FieldErrors = {
  [key: string]: string
}

function getUserErrors(objError: Joi.ValidationResult) {
  const errors: FieldErrors = {}

  if (objError.error) {
    objError.error.details.forEach((err) => {
      errors[err.path.join('.')] = err.message
    })
  }

  return errors
}

export function UserValidate(values: UserForm) {
  const { cpf, email, name, password, surname } = fieldsValidations
  const schema = Joi.object({ cpf, email, name, password, surname })

  return getUserErrors(schema.validate(values, { abortEarly: false }))
}

export function UserValidateSignin(values: UserFormSignIn) {
  const { email, password } = fieldsValidations
  const schema = Joi.object({ email, password })

  return getUserErrors(schema.validate(values, { abortEarly: false }))
}

function getFieldErrors(objError: Joi.ValidationResult, field: keyof UserForm) {
  const errors: FieldErrors = {}

  if (objError.error) {
    objError.error.details.forEach((err) => {
      errors[field] = err.message
    })
  }

  return errors
}

export function FieldValidate(value: string, field: keyof UserForm) {
  return getFieldErrors(
    fieldsValidations[field].validate(value, { abortEarly: false }),
    field
  )
}
