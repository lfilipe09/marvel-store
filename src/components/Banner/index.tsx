import { ArrowRight } from '@styled-icons/feather'
import React from 'react'
import Button from '../../components/Button'
import * as S from './styles'

export type BannerProps = {
  imgUrl: string
  title: string
  subtitle: string
  buttonLabel: string
  buttonLink: string
}

const Banner = ({
  imgUrl,
  title,
  subtitle,
  buttonLabel,
  buttonLink
}: BannerProps) => (
  <>
    <S.Wrapper>
      <S.Caption>
        <S.Title>{title}</S.Title>
        <S.Subtitle dangerouslySetInnerHTML={{ __html: subtitle }} />
        <Button
          color={'white'}
          as="a"
          href={buttonLink}
          size="large"
          icon={<ArrowRight size={'3rem'} strokeWidth={'2px'} />}
        >
          {buttonLabel}
        </Button>
      </S.Caption>
      <S.ImageContainer imgUrl={imgUrl} />
    </S.Wrapper>
  </>
)

export default Banner
