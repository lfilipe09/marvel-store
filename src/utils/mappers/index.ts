import {
  APICharacterComics,
  APIComics,
  APICreatorComics
} from 'types/api/comic-types'
import { convertToSlug } from 'utils/slug'

export const heroBannerMapper = (comicsData: APIComics) => {
  return comicsData.data.results.map((comic) => ({
    imgUrl: comic.thumbnail.path.includes('image_not_available')
      ? 'https://i.ibb.co/gZk8B85/marvel-empty-Prancheta-1.png'
      : `${comic.thumbnail.path}/detail.${comic.thumbnail.extension}`,
    title: comic.title,
    subtitle: `<p>Leia agora <strong>${comic.title}</strong> disponÃ­vel online`,
    buttonLabel: 'Saiba mais',
    buttonLink: `comics/${comic.id}/${convertToSlug(comic.title)}`
  }))
}

export const mainCardsMapper = (comicsData: APIComics) => {
  return comicsData.data.results.map((comic) => ({
    id: comic.id,
    slug: `comics/${comic.id}/${convertToSlug(comic.title)}`,
    imgUrl: comic.thumbnail.path.includes('image_not_available')
      ? 'https://i.ibb.co/gZk8B85/marvel-empty-Prancheta-1.png'
      : `${comic.thumbnail.path}/portrait_fantastic.${comic.thumbnail.extension}`,
    title: comic.title
  }))
}

export const singleComicMapper = (comicsData: APIComics) => {
  return comicsData.data.results.map((comic) => ({
    id: comic.id,
    imgUrl: comic.thumbnail.path.includes('image_not_available')
      ? 'https://i.ibb.co/gZk8B85/marvel-empty-Prancheta-1.png'
      : `${comic.thumbnail.path}/portrait_fantastic.${comic.thumbnail.extension}`,
    title: comic.title,
    description: comic.description,
    variantDescription: comic.variantDescription,
    format: comic.format,
    pageCount: comic.pageCount,
    gallery: comic.images.map((image) => ({
      path: image.path,
      extension: image.extension
    })),
    creators: {
      available: comic.creators.available,
      collectionURI: comic.creators.collectionURI
    },
    characters: {
      available: comic.characters.available,
      collectionURI: comic.characters.collectionURI
    }
  }))
}

export const singleComicCreatorsMapper = (creatorData: APICreatorComics) => {
  return creatorData.data.results.map((creator) => ({
    id: creator.id,
    firstName: creator.firstName,
    lastName: creator.lastName,
    url: creator.urls[0].url
  }))
}

export const singleComicCharacterMapper = (
  characterData: APICharacterComics
) => {
  return characterData.data.results.map((character) => ({
    imgUrl: character.thumbnail.path.includes('image_not_available')
      ? 'https://i.ibb.co/gZk8B85/marvel-empty-Prancheta-1.png'
      : `${character.thumbnail.path}/detail.${character.thumbnail.extension}`,
    name: character.name,
    url: character.urls[0].url
  }))
}
