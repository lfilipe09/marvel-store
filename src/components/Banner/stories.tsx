import React from 'react'
import { Story, Meta } from '@storybook/react/types-6-0'
import Banner, { BannerProps } from '.'

export default {
  title: 'Banner',
  component: Banner,
  args: {
    imgUrl:
      'https://i.annihil.us/u/prod/marvel/i/mg/3/40/4bb4680432f73/detail.jpg',
    title:
      'Marvel Masterworks: The Fantastic Four Vol. V - Variant 2nd Edition (1st) (Trade Paperback)',
    subtitle: '<p>Leia agora <strong>Fantastic Four</strong> dispon√≠vel online',
    buttonLabel: 'Ler agora',
    buttonLink: '/'
  },
  parameters: {
    layout: 'fullscreen'
  }
} as Meta

export const Default: Story<BannerProps> = (args) => (
  <div style={{ maxWidth: '104rem', margin: '0 auto' }}>
    <Banner {...args} />
  </div>
)
