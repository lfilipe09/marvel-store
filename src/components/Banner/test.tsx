import React from 'react'
import { render, screen } from 'utils/test-utils'

import Banner from '.'

const props = {
  imgUrl:
    'https://i.annihil.us/u/prod/marvel/i/mg/3/40/4bb4680432f73/detail.jpg',
  title: 'Defy death',
  subtitle: '<p>Play the new <strong>CrashLands</strong> season',
  buttonLabel: 'Buy now',
  buttonLink: '/games/defy-death'
}

describe('<Banner />', () => {
  it('should render correctly', () => {
    const { container } = render(<Banner {...props} />)

    expect(
      screen.getByRole('heading', { name: /Defy death/i })
    ).toBeInTheDocument()
    expect(
      screen.getByRole('heading', { name: /Play the new CrashLands season/i })
    ).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })
})
