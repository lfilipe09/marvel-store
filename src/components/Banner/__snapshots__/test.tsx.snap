// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Banner /> should render correctly 1`] = `
<main
  class="sc-gsnTZi kYTbhX"
>
  <div
    class="sc-dkzDqf khtZym"
  >
    <h2
      class="sc-hKMtZM joYSWs"
    >
      Defy death
    </h2>
    <h3
      class="sc-eCYdqJ inHoQn"
    >
      <p>
        Play the new 
        <strong>
          CrashLands
        </strong>
         season
      </p>
    </h3>
    <a
      class="sc-bczRLJ jeLxlc"
      color="white"
      href="/games/defy-death"
    >
      <span>
        Buy now
      </span>
      <svg
        aria-hidden="true"
        class="StyledIconBase-ea9ulj-0 lbJwfL"
        fill="none"
        focusable="false"
        height="3rem"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2px"
        viewBox="0 0 24 24"
        width="3rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          x1="5"
          x2="19"
          y1="12"
          y2="12"
        />
        <polyline
          points="12 5 19 12 12 19"
        />
      </svg>
    </a>
  </div>
  <div
    class="sc-jSMfEi bbgzCh"
  />
</main>
`
